ocaml-lsp-server/src/dune.ml

```diff
@@ -1,46 +1,159 @@
 
+module Chan = struct
+  open Fiber.O
+
+  type t =
+    { in_thread : Scheduler.thread
+    ; out_thread : Scheduler.thread
+    ; out_chan : out_channel
+    ; in_chan : in_channel
+    }
+
+  let write t sexp =
+    let+ res =
+      Scheduler.async_exn t.out_thread (fun () ->
+          match sexp with
+          | None -> close_out_noerr t.out_chan
+          | Some sexp -> Csexp.to_channel t.out_chan sexp)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let read t =
+    let+ res =
+      Scheduler.async_exn t.in_thread (fun () ->
+          match Csexp.input_opt t.in_chan with
+          | Error _
+          | Ok None ->
+            Unix.close (Unix.descr_of_in_channel t.in_chan);
+            None
+          | Ok (Some s) -> Some s)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let create scheduler in_fd out_fd =
+    let in_thread = Scheduler.create_thread scheduler in
+    let out_thread = Scheduler.create_thread scheduler in
+    let in_chan = Unix.in_channel_of_descr in_fd in
+    let out_chan = Unix.out_channel_of_descr out_fd in
+    { in_thread; out_thread; in_chan; out_chan }
+end
+
+module Drpc = Dune_rpc.V1
+
+module Client =
+  Drpc.Client
+    (struct
+      include Fiber
+
+      let parallel_iter t ~f =
+        let stream = Fiber.Stream.In.create t in
+        Fiber.Stream.In.parallel_iter stream ~f
+    end)
+    (Chan)
+
 type state =
   | Binary_not_found
   | Out_of_date
-  | Running
 
 type t =
   { scheduler : Scheduler.t
-  ; state : state Fiber.Ivar.t
+  ; diagnostics : Diagnostics.t
   }
 
-let create scheduler = { scheduler; state = Fiber.Ivar.create () }
+let create scheduler diagnostics = { scheduler; diagnostics }
 
-let state t : state Fiber.t =
-  let open Fiber.O in
-  let* state = Fiber.Ivar.peek t.state in
-  match state with
-  | Some s -> Fiber.return s
-  | None -> (
-    match Bin.which "dune" with
-    | None ->
-      let res = Binary_not_found in
-      let+ () = Fiber.Ivar.fill t.state res in
-      res
-    | Some bin ->
-      let bin = Fpath.to_string bin in
-      let dev_null () = Unix.openfile "/dev/null" [ Unix.O_RDWR ] 0o666 in
-      let stdin = dev_null () in
-      let stdout = dev_null () in
-      let stderr = dev_null () in
+let lsp_of_dune dune =
+  let module D = Drpc.Diagnostic in
+  let range =
+    match D.loc dune with
+    | None -> Range.first_line
+    | Some loc ->
+      let loc =
+        { Loc.loc_start = loc.start; loc_end = loc.stop; loc_ghost = false }
+      in
+      Range.of_loc loc
+  in
+  let severity =
+    D.severity dune
+    |> Option.map ~f:(function
+         | D.Error -> DiagnosticSeverity.Error
+         | Warning -> DiagnosticSeverity.Warning)
+  in
+  let message = Format.asprintf "%a@." Pp.to_fmt (D.message dune) in
+  Diagnostic.create ~range ?severity ~source:"dune" ~message ()
```

Comment:
This function is surprisingly small considering that it's the bridge between Dune RPC and LSP! This is great 🎉

Reply:
Thanks! Hopefully this paves the way for adding dune rpc features to vscode as well.